File: src/backend/lyrics/Spotify.ts
import { TOTP } from "totp-generator"
import { loadCredentials } from "../Credentials"
import { GenericLyrics, GenericLyricsLine } from "../../common/types/LyricsTypes"

const SPOTIFY_WEB_TOKEN_URL = 'https://open.spotify.com/api/token'
const SPOTIFY_CLIENT_TOKEN_URL = 'https://clienttoken.spotify.com/v1/clienttoken'
const SPOTIFY_LYRICS_BASE_URL = 'https://spclient.wg.spotify.com/color-lyrics/v2/track'
const SPOTIFY_USER_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36'

// Hats off to this guy for hosting the secrets file :)
const SECRETS_URL = 'https://git.gay/thereallo/totp-secrets/raw/branch/main/secrets/secrets.json'

const SPOTIFY_APP_VERSION = "1.2.75.277.g5c44d208"

type SpotifyLyrics = {
    lyrics: {
        syncType: 'LINE_SYNCED' | 'LINE_UNSYNCED',
        lines: SpotifyLyricsLine[],
        provider: 'MusixMatch',
        providerLyricsId: number,
        providerDisplayName: 'Musixmatch',
        syncLyricsUri: string,
        isDenseTypeface: boolean,
        alternatives: [],
        langauge: string,
        isRtlLanguage: boolean,
        capStatus: string,
        previewLines: []
    },
    colors: {
        background: number,
        text: number,
        highlightText: number,
    },
    hasVocalRemoval: false
}

type SpotifyLyricsLine = {
    startTimeMs: string,
    words: string,
    syllables: [],
    endTimeMs: string,
    transliteratedWords: string
}

type SpotifySecret = {
    version: number,
    secret: string
}

type SpotifyAccessTokenResponse = {
    clientId: string,
    accessToken: string,
    accessTokenExpirationTimestampMs: number,
    isAnonymous: true,
    _notes: string,
}

type SpotifyClientTokenResponse = {
    response_type: string,
    granted_token: {
        token: string,
        expires_after_seconds: number,
        refresh_after_seconds: number,
        domains: unknown
    }
}

/* 
Credits to the guys at this discussion: https://github.com/librespot-org/librespot/discussions/1562
for helping figure out how spotify's TOTP generation works.
Also thanks to the guys at https://github.com/KRTirtho/spotube for actually implementing it in their code,
since I mostly just adapted it from there :P
*/

const cleanBuffer = (e: string): Uint8Array => {
    e = e.replaceAll(" ", "");
    const length = Math.floor(e.length / 2);
    const n = new Uint8Array(length);
    for (let r = 0; r < e.length; r += 2) {
      n[Math.floor(r / 2)] = parseInt(e.substring(r, r + 2), 16);
    }
    return n;
}

const base32FromBytes = (e: Uint8Array, secretSauce: string): string => {
    let t = 0;
    let n = 0;
    let r = "";
    for (let i = 0; i < e.length; i++) {
      n = n << 8 | e[i];
      t += 8;
      while (t >= 5) {
        r += secretSauce[(n >>> (t - 5)) & 31];
        t -= 5;
      }
    }
    if (t > 0) {
      r += secretSauce[(n << (5 - t)) & 31];
    }
    return r;
}

const generateTotp = async () => {
    const secretSauce = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"

    const secrets = await (await fetch(SECRETS_URL)).json()
    const mostRecentSecret: SpotifySecret = secrets[secrets.length - 1]
    console.log("Most recent secret:", mostRecentSecret)

    const secretArray = mostRecentSecret.secret
		.toString()
		.split("")
		.map((char: string) => char.charCodeAt(0));
	const secretCipherBytes = secretArray.map((e: number, t: number) => e ^ ((t % 33) + 9));

	const secretBytes = cleanBuffer(
		new TextEncoder()
			.encode(secretCipherBytes.join(""))
			.reduce((acc, val) => acc + val.toString(16).padStart(2, "0"), ""),
	);

	const secret = base32FromBytes(secretBytes, secretSauce);

	return {
        totp: (await TOTP.generate(secret)).otp,
        version: mostRecentSecret.version
    };
}

/**
 * Spotify lyrics provider using Spotify's internal (mostly undocumented) API.
 */
class SpotifyProvider {
    private initPromise: Promise<void>
    private CLIENT_TOKEN: string
    private ACCESS_TOKEN: string

    constructor() {
        this.initPromise = this.initialize()
    }

    // Initialize the provider by fetching necessary tokens.
    // I swear, spotify is so fucking annoying with this, jesus christ.
    // Like, why can't you just have a public lyrics API, Spotify?
    // This will most likely break in the future, but whatever, what can you do.
    private async initialize() {
        const {totp, version} = await generateTotp()
        const cookie = (await loadCredentials()).spotifyCookie
        if (!cookie) throw new Error("Spotify cookie not found in credentials.")

        const ACCESS_TOKEN_URL = `${SPOTIFY_WEB_TOKEN_URL}?reason=init&productType=web-player&totp=${totp}&totpServer=${totp}&totpVer=${version}`

        // Use user's provided spotify to generate a valid access token. 

        const res = await fetch(ACCESS_TOKEN_URL, {
            headers: {
                'Cookie': cookie,
                'User-Agent': SPOTIFY_USER_AGENT
            }
        })
        
        const accessTokenResponse: SpotifyAccessTokenResponse = await res.json()
        this.ACCESS_TOKEN = accessTokenResponse.accessToken

        // Generate random device ID (32-character hex string), so the user stays anonymous.
        // This is probably not necessary, but eh, why not.
        // Also, spotify seems to only accept lowercase hex characters, so we do that too.
        const deviceId = Array.from({length: 32}, () => 
            Math.floor(Math.random() * 16).toString(16)
        ).join('')
        
        const data = {
            client_data: {
                client_id: accessTokenResponse.clientId,
                client_version: SPOTIFY_APP_VERSION,
                js_sdk_data: {
                    device_brand: "unknown",
                    device_id: deviceId,
                    device_model: "unknown",
                    device_type: "computer",
                    os: "windows",
                    os_version: "NT 10.0"
                }
            }
        }

        // Fetch client token, which is required for the lyrics API.
        // Should last for a long time, so we only fetch it once during initialization.
        // Also, fetch a new one every time we initialize the provider (AKA when the app starts).
        // Spotify really loves their tokens, huh.
        const clientTokenRes = await fetch(SPOTIFY_CLIENT_TOKEN_URL, {
            method: "POST",
            body: JSON.stringify(data),
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
        })
        
        const clientTokenResponse: SpotifyClientTokenResponse = await clientTokenRes.json()
        if (clientTokenResponse.response_type == "RESPONSE_GRANTED_TOKEN_RESPONSE") {
            this.CLIENT_TOKEN = clientTokenResponse.granted_token.token
        }
    }

    // Thank GOD, we can now finally fetch the lyrics.
    async getLyrics(id: string): Promise<GenericLyrics | null> {
        // Wait for initialization to complete
        await this.initPromise
        
        const URL = `${SPOTIFY_LYRICS_BASE_URL}/${id}?format=json&vocalRemoval=false&market=from_token`

        console.log("Client Token:", this.CLIENT_TOKEN)
        console.log("Access Token:", this.ACCESS_TOKEN)
        
        const response = await fetch(URL, {
            // Necessary headers for Spotify Closed API response.
            // Seriously Spotify? You guys couldn't expose this?
            // Holy, these guys are something else.
            headers: {
                'App-Platform': 'WebPlayer',
                'Accept': 'application/json',
                'Authorization': `Bearer ${this.ACCESS_TOKEN}`,
                'Client-Token': this.CLIENT_TOKEN,
                'User-Agent': SPOTIFY_USER_AGENT,
                'Spotify-App-Version': SPOTIFY_APP_VERSION,
            },
        })

        try {
            if (response.ok) {
                const lyricsResponse = (await response.json()) as SpotifyLyrics
                const synced = lyricsResponse.lyrics.syncType == "LINE_SYNCED" ? true : false
                const lines: GenericLyricsLine[] = lyricsResponse.lyrics.lines.map((line) => {
                    return {
                        content: line.words,
                        startTimeMs: line.startTimeMs
                    }
                })
                const properLyrics: GenericLyrics = {
                    provider: "spotify",
                    synced: synced,
                    lines: lines
                }

                return properLyrics
            } else if (response.status === 404) {
                console.log("No lyrics found for this track on Spotify.")
                return null
            } else {
                console.error("Error fetching lyrics from Spotify:", response.status, response.statusText)
                return null
            }
        } catch (e) {
            console.log(e)
        }
    }
}

export const spotifyProvider = new SpotifyProvider()

File: lib/providers/lyrics/spotify.dart
/// Spotify lyrics provider using internal API + TOTP
library;

import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:typed_data';
import 'package:crypto/crypto.dart';
import 'package:http/http.dart' as http;
import 'package:http/io_client.dart';
import '../../models/metadata_models.dart';
import '../../services/credentials.dart';
import '../../utils/logger.dart';

const _spotifyWebTokenUrl = 'https://open.spotify.com/api/token';
const _spotifyClientTokenUrl = 'https://clienttoken.spotify.com/v1/clienttoken';
const _spotifyLyricsBaseUrl =
    'https://spclient.wg.spotify.com/color-lyrics/v2/track';
const _spotifyUserAgent =
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
    '(KHTML, like Gecko) Chrome/141.0.0.0 Safari/537.36';
const _spotifyAppVersion = '1.2.83.373.ge7c77344';
const _allowInsecureSpotifyTls = bool.fromEnvironment(
  'WISP_ALLOW_INSECURE_SPOTIFY_TLS',
  defaultValue: true,
);
const _allowInsecureSpotifySecretsTls = bool.fromEnvironment(
  'WISP_ALLOW_INSECURE_SPOTIFY_SECRETS_TLS',
  defaultValue: true,
);

// Hosts TOTP secrets
const _secretsUrl =
    'https://git.gay/thereallo/totp-secrets/raw/branch/main/secrets/secrets.json';

class SpotifyLyricsProvider {
  final CredentialsService _credentialsService = CredentialsService();
  Future<void>? _initFuture;
  String? _clientToken;
  String? _accessToken;

  Future<void> _ensureInitialized() {
    _initFuture ??= _initialize();
    return _initFuture!;
  }

  Future<void> _initialize() async {
    try {
      logger.i("Initializing Spotify lyrics provider...");

      logger.i("Fetching Spotify lyrics cookie...");

      final cookie = await _credentialsService.getSpotifyLyricsCookie();
      if (cookie == null || cookie.isEmpty) {
        throw StateError('Spotify lyrics cookie (sp_dc) not configured');
      }

      logger.i("Got Spotify cookie, requesting access token...");

      final accessJson = await _requestAccessToken(cookie);
      final accessToken = accessJson['accessToken'] as String?;
      final clientId = accessJson['clientId'] as String?;
      if (accessToken == null || clientId == null) {
        throw StateError('Invalid access token response');
      }

      logger.i("Access token obtained, requesting client token...");

      _accessToken = accessToken;

      final deviceId = _randomHex(32);
      final clientTokenPayload = {
        'client_data': {
          'client_id': clientId,
          'client_version': _spotifyAppVersion,
          'js_sdk_data': {
            'device_brand': 'unknown',
            'device_id': deviceId,
            'device_model': 'unknown',
            'device_type': 'computer',
            'os': 'windows',
            'os_version': 'NT 10.0',
          },
        },
      };

      final client = _createSpotifyHttpClient();
      http.Response clientTokenResponse;
      try {
        clientTokenResponse = await client.post(
          Uri.parse(_spotifyClientTokenUrl),
          headers: {
            'Content-Type': 'application/json',
            'Accept': 'application/json',
          },
          body: jsonEncode(clientTokenPayload),
        );
      } finally {
        client.close();
      }

      if (clientTokenResponse.statusCode != 200) {
        throw StateError(
          'Spotify client token request failed: '
          '${clientTokenResponse.statusCode}',
        );
      }

      logger.i(
        "Client token obtained, Spotify lyrics provider initialized successfully",
      );

      final clientJson =
          jsonDecode(clientTokenResponse.body) as Map<String, dynamic>;
      final responseType = clientJson['response_type'] as String?;
      if (responseType == 'RESPONSE_GRANTED_TOKEN_RESPONSE') {
        final grantedToken =
            (clientJson['granted_token'] as Map<String, dynamic>?)?['token']
                as String?;
        _clientToken = grantedToken;
      }

      if (_clientToken == null || _clientToken!.isEmpty) {
        throw StateError('Invalid client token response');
      }
    } catch (e, stackTrace) {
      _initFuture = null;
      logger.e(
        'Failed to initialize Spotify lyrics provider',
        error: e,
        stackTrace: stackTrace,
      );
      rethrow;
    }
  }

  Future<LyricsResult?> getLyrics(String trackId) async {
    try {
      await _ensureInitialized();
    } catch (_) {
      return null;
    }

    if (_accessToken == null || _clientToken == null) return null;

    final normalizedId = _normalizeTrackId(trackId);
    if (normalizedId.isEmpty) return null;

    final url = Uri.parse('$_spotifyLyricsBaseUrl/$normalizedId').replace(
      queryParameters: {
        'format': 'json',
        'vocalRemoval': 'false',
        'market': 'from_token',
      },
    );

    final client = _createSpotifyHttpClient();
    http.Response response;
    try {
      response = await client.get(
        url,
        headers: {
          'App-Platform': 'WebPlayer',
          'Accept': 'application/json',
          'Authorization': 'Bearer $_accessToken',
          'Client-Token': _clientToken!,
          'User-Agent': _spotifyUserAgent,
          'Spotify-App-Version': _spotifyAppVersion,
        },
      );
    } finally {
      client.close();
    }

    if (response.statusCode == 404) return null;
    if (response.statusCode != 200) {
      logger.w('Spotify lyrics request failed: ${response.statusCode}');
      return null;
    }

    try {
      final json = jsonDecode(response.body) as Map<String, dynamic>;
      final lyrics = json['lyrics'] as Map<String, dynamic>?;
      if (lyrics == null) return null;
      final syncType = lyrics['syncType'] as String? ?? 'LINE_UNSYNCED';
      final linesJson = lyrics['lines'] as List? ?? const [];
      final lines = linesJson
          .whereType<Map<String, dynamic>>()
          .map((line) {
            final content = (line['words'] as String?)?.trim() ?? '';
            final startMsRaw = line['startTimeMs'];
            final startTimeMs = startMsRaw is String
                ? int.tryParse(startMsRaw) ?? 0
                : startMsRaw is int
                ? startMsRaw
                : 0;
            return LyricsLine(content: content, startTimeMs: startTimeMs);
          })
          .where((line) => line.content.isNotEmpty)
          .toList();

      return LyricsResult(
        provider: LyricsProviderType.spotify,
        synced: syncType == 'LINE_SYNCED',
        lines: lines,
      );
    } catch (e, stackTrace) {
      logger.e(
        'Failed to parse Spotify lyrics response',
        error: e,
        stackTrace: stackTrace,
      );
      return null;
    }
  }
}

class _TotpPayload {
  final String otp;
  final int version;

  const _TotpPayload({required this.otp, required this.version});
}

Future<Map<String, dynamic>> _requestAccessToken(String cookie) async {
  Future<Map<String, dynamic>> fetchWithTotp(_TotpPayload totpPayload) async {
    final accessTokenUrl = Uri.parse(_spotifyWebTokenUrl).replace(
      queryParameters: {
        'reason': 'init',
        'productType': 'web-player',
        'totp': totpPayload.otp,
        'totpServer': totpPayload.otp,
        'totpVer': totpPayload.version.toString(),
      },
    );

    final client = _createSpotifyHttpClient();
    http.Response accessTokenResponse;
    try {
      accessTokenResponse = await client.get(
        accessTokenUrl,
        headers: {
          'Cookie': _normalizeCookie(cookie),
          'User-Agent': _spotifyUserAgent,
          'Accept': 'application/json',
          'Origin': 'https://open.spotify.com',
          'Referer': 'https://open.spotify.com/',
        },
      );
    } finally {
      client.close();
    }

    if (accessTokenResponse.statusCode != 200) {
      final body = accessTokenResponse.body;
      final snippet = body.length > 300 ? body.substring(0, 300) : body;
      throw StateError(
        'Spotify access token request failed: '
        '${accessTokenResponse.statusCode} ${snippet.isEmpty ? '' : snippet}',
      );
    }

    return jsonDecode(accessTokenResponse.body) as Map<String, dynamic>;
  }

  final firstTotp = await _generateTotp();
  try {
    return await fetchWithTotp(firstTotp);
  } catch (e) {
    final message = e.toString();
    if (!message.contains('400')) rethrow;
  }

  final retryTotp = await _generateTotp();
  return fetchWithTotp(retryTotp);
}

Future<_TotpPayload> _generateTotp() async {
  const secretSauce = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

  final client = _createHttpClient(
    allowInsecureSpotifySecretsTls: _allowInsecureSpotifySecretsTls,
  );
  http.Response response;
  try {
    response = await client.get(Uri.parse(_secretsUrl));
  } finally {
    client.close();
  }
  if (response.statusCode != 200) {
    throw StateError('Failed to fetch TOTP secrets');
  }

  final secrets = jsonDecode(response.body) as List<dynamic>;
  if (secrets.isEmpty) {
    throw StateError('No secrets available for TOTP');
  }

  final mostRecent = secrets.last as Map<String, dynamic>;
  final version = mostRecent['version'] as int? ?? 0;
  final secretValue = mostRecent['secret'] as String? ?? '';
  if (secretValue.isEmpty) {
    throw StateError('Invalid TOTP secret payload');
  }

  final secretArray = secretValue.codeUnits;
  final secretCipherBytes = <int>[];
  for (var i = 0; i < secretArray.length; i++) {
    secretCipherBytes.add(secretArray[i] ^ ((i % 33) + 9));
  }

  final cipherString = secretCipherBytes.join('');
  final cipherBytes = utf8.encode(cipherString);
  final hexBuffer = StringBuffer();
  for (final value in cipherBytes) {
    hexBuffer.write(value.toRadixString(16).padLeft(2, '0'));
  }

  final secretBytes = _cleanBuffer(hexBuffer.toString());
  final base32Secret = _base32FromBytes(secretBytes, secretSauce);

  final otp = _generateTotpCode(base32Secret);
  return _TotpPayload(otp: otp, version: version);
}

Uint8List _cleanBuffer(String hex) {
  final sanitized = hex.replaceAll(' ', '');
  final length = sanitized.length ~/ 2;
  final bytes = Uint8List(length);
  for (var i = 0; i < sanitized.length; i += 2) {
    bytes[i ~/ 2] = int.parse(sanitized.substring(i, i + 2), radix: 16);
  }
  return bytes;
}

String _base32FromBytes(Uint8List bytes, String alphabet) {
  var t = 0;
  var n = 0;
  final buffer = StringBuffer();
  for (var i = 0; i < bytes.length; i++) {
    n = (n << 8) | bytes[i];
    t += 8;
    while (t >= 5) {
      buffer.write(alphabet[(n >> (t - 5)) & 31]);
      t -= 5;
    }
  }
  if (t > 0) {
    buffer.write(alphabet[(n << (5 - t)) & 31]);
  }
  return buffer.toString();
}

String _generateTotpCode(
  String base32Secret, {
  int digits = 6,
  int interval = 30,
}) {
  final secretBytes = _decodeBase32(base32Secret);
  final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
  final counter = timestamp ~/ interval;
  final counterBytes = ByteData(8)..setInt64(0, counter);

  final hmac = Hmac(sha1, secretBytes);
  final digest = hmac.convert(counterBytes.buffer.asUint8List()).bytes;
  final offset = digest.last & 0x0f;
  final code =
      ((digest[offset] & 0x7f) << 24) |
      ((digest[offset + 1] & 0xff) << 16) |
      ((digest[offset + 2] & 0xff) << 8) |
      (digest[offset + 3] & 0xff);
  final mod = pow(10, digits).toInt();
  final otp = code % mod;
  return otp.toString().padLeft(digits, '0');
}

Uint8List _decodeBase32(String input) {
  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
  final normalized = input.replaceAll('=', '').toUpperCase();
  var buffer = 0;
  var bitsLeft = 0;
  final bytes = <int>[];

  for (final char in normalized.codeUnits) {
    final index = alphabet.indexOf(String.fromCharCode(char));
    if (index < 0) continue;
    buffer = (buffer << 5) | index;
    bitsLeft += 5;
    if (bitsLeft >= 8) {
      bitsLeft -= 8;
      bytes.add((buffer >> bitsLeft) & 0xff);
    }
  }

  return Uint8List.fromList(bytes);
}

String _randomHex(int length) {
  const hex = '0123456789abcdef';
  final rand = Random.secure();
  final buffer = StringBuffer();
  for (var i = 0; i < length; i++) {
    buffer.write(hex[rand.nextInt(16)]);
  }
  return buffer.toString();
}

String _normalizeTrackId(String trackId) {
  var id = trackId.trim();
  if (id.contains('?')) {
    id = id.split('?').first;
  }
  if (id.contains('spotify:')) {
    final parts = id.split(':');
    id = parts.isNotEmpty ? parts.last : id;
  }
  if (id.contains('/')) {
    final uri = Uri.tryParse(id);
    if (uri != null && uri.pathSegments.isNotEmpty) {
      id = uri.pathSegments.last;
    }
  }
  return id.trim();
}

String _normalizeCookie(String cookie) {
  final trimmed = cookie.trim();
  if (trimmed.startsWith('sp_dc=')) return trimmed;
  return 'sp_dc=$trimmed';
}

http.Client _createHttpClient({required bool allowInsecureSpotifySecretsTls}) {
  if (!allowInsecureSpotifySecretsTls) {
    return http.Client();
  }

  final ioClient = HttpClient();
  ioClient.badCertificateCallback = (cert, host, port) {
    return host == 'git.gay';
  };
  return IOClient(ioClient);
}

http.Client _createSpotifyHttpClient() {
  if (!_allowInsecureSpotifyTls) {
    return http.Client();
  }

  final ioClient = HttpClient();
  ioClient.badCertificateCallback = (cert, host, port) {
    return host == 'open.spotify.com' ||
        host == 'clienttoken.spotify.com' ||
        host == 'spclient.wg.spotify.com';
  };
  return IOClient(ioClient);
}
